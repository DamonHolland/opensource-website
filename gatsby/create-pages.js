const path = require(`path`);

const allProjectsQuery = `
  query allProjects {
    allProjects(filter: {projectType: {eq: "newrelic"}}) {
      edges {
        node {
          ...exploreProjectsFields
        }
      }
    }
    allCategories: allProjects {
      group(field: ossCategory___title) {
        fieldValue
        totalCount
      }
    }
    allLanguages: allProjects {
      group(field: stats___languages___name) {
        fieldValue
        totalCount
      }
    }
    
    allProjectTags {
      group(field: title) {
        fieldValue
      }
    }
  }

  fragment exploreProjectsFields on Projects {
    id
    slug
    fullName
    githubUrl
    stats {
      id
      contributors
      commits
      languages {
        id
        name
        color
      }
      releases
      screenshots
    }
    description
    iconUrl
    name
    ossCategory {
      slug
      title
      description
    }
    owner {
      login
      type
    }
    permalink
    primaryLanguage
    projectTags {
      slug
      title
    }
    shortDescription
    supportUrl
    projectTags {
      title
      slug
    }
    title
    version
    website {
      title
      url
    }
  }
`;

const getProjectComponent = projectType => {
  if (projectType === 'external') {
    return path.resolve('./src/templates/external-project-page.js');
  }

  if (projectType === 'external') {
    return path.resolve('./src/templates/partner-project-page.js');
  }

  return path.resolve(`./src/templates/project-page.js`);
};

const createProjectPages = async ({ graphql, actions }) => {
  const { createPage } = actions;
  const result = await graphql(`
    query AllProjects {
      allProjects {
        edges {
          node {
            id
            name
            slug
            fullName
            permalink
            projectType
          }
        }
      }
    }
  `);
  const pages = result.data.allProjects.edges;

  pages.forEach(({ node }) => {
    const component = getProjectComponent(node.projectType);

    // https://www.gatsbyjs.org/docs/actions/#createPage
    createPage({
      path: `projects/${node.fullName}`, // {org/user}/{repository-foo-name}, ex. newrelic/nr1-workload-geoops
      component,
      context: {
        // Data passed to context is available
        // in page queries as GraphQL variables.

        slug: node.slug // {org/user}-{repository-foo-name}, ex. newrelic-nr1-workload-geoops
      }
    });
  });
};

/*
 * Note:
 *    Although our exploreProjectsFields fragment defined in src/fragments/index.js is available in /pages,
 *    it is not automagically available when running graphql here. For now, we've copied it.
 *
 * TO DO:
 *    Investigate if Gatsby has this available somewhere when running the API methods, perhaps
 *    something like graphql.fragments[name of fragment]
 */
const createExploreProjects = async ({ graphql, actions }) => {
  const { createPage } = actions;

  const allProjectsResult = await graphql(allProjectsQuery);
  const {
    allProjects,
    allCategories,
    allLanguages,
    allProjectTags
  } = allProjectsResult.data;

  createPage({
    path: '/explore-projects',
    component: path.resolve(`./src/templates/explore-projects.js`),
    context: {
      projectData: {
        allProjects: allProjects.edges.map(p => p.node),
        allLanguages: allLanguages.group,
        allCategories: allCategories.group,
        allProjectTags: allProjectTags.group,
        // sortBy:
        searchEngineOptions: {
          indexStrategy: 'Prefix match',
          searchSanitizer: 'Lower Case',
          indexFields: [
            ['slug'],
            ['fullName'],
            ['name'],
            ['description'],
            ['ossCategory', 'title'],
            ['website', 'title'],
            ['projectTags', 'title'],
            ['primaryLanguage']

            // Nested array syntax doesn't work until this PR is merged:
            // https://github.com/bvaughn/js-search/pull/78/files#diff-5d56a676b2913ba26d67295fe642d9b0R18
            // ['stats', 'languages', '[]', 'name']
          ],
          removeStopWords: true
        }
      }
    }
  });
};

module.exports = async params => {
  createProjectPages(params);
  createExploreProjects(params);
};
