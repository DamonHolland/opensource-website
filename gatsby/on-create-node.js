// https://www.gatsbyjs.org/docs/actions/#createNode

const { v4: uuidv4 } = require('uuid');

const isMdx = type => type === 'Mdx';
const slugFromAbsoluteFilePath = (dir, fileAbsolutePath) =>
  fileAbsolutePath
    .slice(fileAbsolutePath.indexOf(dir) + dir.length, fileAbsolutePath.length)
    .replace('.mdx', '');

const createProjectMainContent = ({ node, actions }) => {
  const { createNode } = actions;
  const projectMainContentDir = 'src/data/project-main-content/';
  const isMainContent =
    node.fileAbsolutePath.indexOf(projectMainContentDir) >= 0;

  if (!isMainContent) {
    return;
  }

  const fileAbsolutePath = node.fileAbsolutePath;
  const slug = slugFromAbsoluteFilePath(
    projectMainContentDir,
    fileAbsolutePath
  );

  const fieldData = {
    slug
  };
  createNode({
    ...fieldData,

    // Required fields.
    id: uuidv4(),
    parent: node.id,
    children: [],
    internal: {
      type: `ProjectMainContent`,
      contentDigest: node.internal.contentDigest // We can safely reuse this hash since it's content-based
    }
  });
};

module.exports = params => {
  const { node } = params;

  // if (node.internal.type === `projectsJson`) {
  //   const basePath = `projects`;
  //   const slug = createFilePath({ node, getNode, basePath });
  //   createNodeField({
  //     node,
  //     name: `slug`,
  //     value: slug
  //   });
  // }

  // Transform readmes into an object type of ProjectMainContent which we'll stitch into GraphQL in `type-defs.gql`
  // This will enable us to get all data needed for a page with one dynamic query
  if (isMdx(node.internal.type)) {
    createProjectMainContent(params);
  }
};
