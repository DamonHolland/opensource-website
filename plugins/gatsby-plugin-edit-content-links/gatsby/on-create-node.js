// https://www.gatsbyjs.org/docs/actions/#createNode

// TO DO - allow input parameters from gatsby-config.js
const githubUrl = () => {
  const options = {
    host: 'https://github.com',
    org: 'newrelic',
    repo: 'opensource-website',
    defaultBranch: 'develop'
  };

  const { host, org, repo, defaultBranch } = options;
  const githubBaseUrl = [host, org, repo, 'tree', defaultBranch].join('/');
  return githubBaseUrl;
};

const editLinksMap = {
  '/projects/(.*)': `src/data/project-main-content/$match.mdx`,
  '^/$': `src/pages/index.js`
};

const getEditLinkFromPath = ({ path }) => {
  const result = Object.entries(editLinksMap).reduce(
    (previousValue, [expression, value]) => {
      if (previousValue.link) {
        return previousValue;
      }

      const match = path.match(expression);

      if (Array.isArray(match) && match[0]) {
        if (value.indexOf('$match') >= 0) {
          return { link: value, slug: match[1] };
        }
        return { link: value, slug: '' };
      }

      return previousValue;
    },
    { link: false, slug: '' }
  );

  if (result.link) {
    if (result.link.indexOf('$match') >= 0) {
      const link = result.link.replace('$match', result.slug.replace('/', '-'));
      return link;
    }
    const link = result.link;
    return link;
  }
};

module.exports = params => {
  const { node, actions } = params;
  const { createNodeField } = actions;
  const { path, componentPath } = node;
  const cwd = process.cwd();

  if (node.internal.type === 'SitePage') {
    if (!path || !componentPath) {
      // console.log(JSON.stringify(node, null, 2));
      return;
    }
    const githubBaseUrl = githubUrl();
    const mappedEditLink = getEditLinkFromPath({
      path
    });

    const sourceFilePath = componentPath.replace(`${cwd}/`, '');
    const editLink = !mappedEditLink ? sourceFilePath : mappedEditLink;

    const fullUrl = [githubBaseUrl, editLink].join('/');
    // console.log(fullUrl);

    createNodeField({
      node,
      name: `contentEditLink`,
      value: fullUrl
    });
  }
};
