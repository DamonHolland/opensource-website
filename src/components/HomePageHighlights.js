import React from 'react';
import PropTypes from 'prop-types';

import styles from './HomePageHighlights.module.scss';

const HomePageHighlights = data => {
  const projects = data.data;

  return (
    <div className={styles.featuredExternalProjects}>
      {projects.map(project => {
        return (
          <div className={styles.featuredExternalProject} key={project.title}>
            <img
              className={styles.featuredExternalProjectIcon}
              src={project.icon}
              alt={`icon for ${project.title}`}
            />
            <h4 className={styles.featuredExternalProjectTitle}>
              {project.title}
            </h4>
            <p className={styles.featuredExternalProjectDescription}>
              {project.description}
            </p>
            <a
              href="#"
              className={`button ${styles.featuredExternalProjectButton}`}
            >
              Read more
            </a>
            <footer className={styles.featuredExternalProjectFooter}>
              <a
                className={styles.featuredExternalProjectFooterLink}
                href={project.githubUrl}
              >
                View in Github
              </a>
              <a
                className={styles.featuredExternalProjectFooterLink}
                href={project.website}
              >
                Go to website
              </a>
            </footer>
          </div>
        );
      })}
    </div>
  );
};

/* eslint-disable react/no-unused-prop-types */
HomePageHighlights.propTypes = {
  data: PropTypes.array
};
/* eslint-enable */

export default HomePageHighlights;
