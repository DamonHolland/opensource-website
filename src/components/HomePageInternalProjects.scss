.featured-internal-projects {
  display: grid;
  grid-template-columns: 1fr 1fr 1fr 1fr;
  grid-template-rows: 1frÂ  auto;
  grid-gap: 7.941176471% 1.694915254237288%;
}

.featured-internal-project {
  display: flex;
  align-items: center;
  padding: 16px 20px;
  box-sizing: border-box;
  background: #FFFFFF;
  border: 1px solid var(--color-neutrals-300);
  box-sizing: border-box;
  box-shadow: 0px 0.249053px 0.553451px rgba(0, 0, 0, 0.00562291), 0px 0.598509px 1.33002px rgba(0, 0, 0, 0.00807786), 0px 1.12694px 2.50431px rgba(0, 0, 0, 0.01), 0px 2.01027px 4.46726px rgba(0, 0, 0, 0.0119221), 0px 3.75998px 8.35552px rgba(0, 0, 0, 0.0143771), 0px 9px 20px rgba(0, 0, 0, 0.02);
  border-radius: 4px;
  transition: all .1s var(--ease-out-quad);

  &:hover {
    transform: translateY(-2px);
    border: 1px solid var(--color-neutrals-400);
  }
  
  &:active {
    transform: translateY(2px);
  }
}

.featured-internal-project-icon {
  display: block;
  width: 55px;
  height: 55px;
  margin-right: 14px;
  border-radius: 50%;
}

.featured-internal-project-title {
  margin-top: 0;
  margin-bottom: 2px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  max-width: 19ch;
}

.featured-internal-project-description {
  display: -webkit-box;
  margin-top: 0;
  margin-bottom: 4px;
  -webkit-line-clamp: 2;
  -webkit-box-orient: vertical;
  overflow: hidden;
  font-size: 14px;
  line-height: 20px;
  color: var(--color-neutrals-600);
}