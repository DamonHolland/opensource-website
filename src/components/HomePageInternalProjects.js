
import React from 'react'
import PropTypes from 'prop-types'
import { Link } from "gatsby"

import styles from './HomePageInternalProjects.module.scss'

const HomePageInternalProjects = (data) => {
  const projects = data.data

  return (
    <div className={styles.featuredInternalProjects}>
      {projects.map(project => {
        return (
          <Link to={project.permalink} className={styles.featuredInternalProject} key={project.title}>
            <img src={project.iconUrl} className={styles.featuredInternalProjectIcon} alt={`Icon for ${project.title}`}/>
            <div className={styles.featuredInternalProjectMeta}>
              <h4 className={styles.featuredInternalProjectTitle}>{project.title}</h4>
              <p className={styles.featuredInternalProjectDescription}>{project.shortDescription}</p>
            </div>
          </Link> 
        )
      })}
    </div>
  )
}

HomePageInternalProjects.propTypes = {
  data: PropTypes.array
}

export default HomePageInternalProjects