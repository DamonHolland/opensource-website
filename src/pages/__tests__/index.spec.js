import React from 'react';
import TestRenderer from 'react-test-renderer'; // https://reactjs.org/docs/test-renderer.html
import { render } from '@testing-library/react'; // https://testing-library.com/docs/react-testing-library/intro
import { useStaticQuery } from 'gatsby';

import HomePage from '../index';
import HomePageHighlights from '../../components/HomePageHighlights';
import HomePageInternalProjects from '../../components/HomePageInternalProjects';

beforeEach(() => {
  useStaticQuery.mockImplementationOnce(() => ({
    site: {
      siteMetadata: {
        title: 'New Relic Open Source',
        description:
          "New Relic <3's open source. We built this site to make it easy for you to explore the projects we're maintaining as well as our involvement in open standards.",
        author: '@newrelic'
      }
    }
  }));
});

describe('HomePage', () => {
  it('Renders correctly', () => {
    const tree = TestRenderer.create(<HomePage />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('HomePage Hero Header', () => {
  it('Displays the correct title', () => {
    const { getByTestId } = render(<HomePage />);
    const expectation = getByTestId('homepageHeroHeading');
    expect(expectation).toHaveTextContent(
      'Open standards. Open instrumentation. Open collaboration.'
    );
  });
});

describe('HomePage Projects We Support', () => {
  it('Renders correctly', () => {
    const testRenderer = TestRenderer.create(<HomePage />);
    const testInstance = testRenderer.root;
    const props = testInstance.findByType(HomePageHighlights).props;

    expect(Array.isArray(props.data)).toBe(true);
  });
});

describe('HomePage Explore Projects', () => {
  it('renders correctly', () => {
    const testRenderer = TestRenderer.create(<HomePage />);
    const testInstance = testRenderer.root;

    const props = testInstance.findByType(HomePageInternalProjects).props;

    expect(Array.isArray(props.data)).toBe(true);
  });
});
