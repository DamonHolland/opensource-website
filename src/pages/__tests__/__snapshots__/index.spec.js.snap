// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage Renders correctly 1`] = `
<div
  className="layout-container undefined"
>
  <div
    className="globalHeaderContainer  "
  >
    <div
      className="globalHeaderContent"
    >
      <div
        className="leftSideLinksContainer"
      >
        <a
          className="logo"
          href="https://newrelic.com/"
          rel="noopener noreferrer"
        >
          New Relic
        </a>
        <ul
          className="leftSideLinks"
        >
          <li
            className="leftSideLinkItem"
          >
            <a
              className="leftSideLink"
              href="https://developer.newrelic.com/"
            >
              Developers
            </a>
          </li>
          <li
            className="leftSideLinkItem active"
          >
            <a
              className="leftSideLink"
              href="#"
            >
              Open Source
            </a>
          </li>
          <li
            className="leftSideLinkItem"
          >
            <a
              className="leftSideLink"
              href="https://docs.newrelic.com/"
            >
              Docs
            </a>
          </li>
        </ul>
      </div>
      <ul
        className="rightSideButtons"
      >
        <li
          className="darkModeToggle"
        />
        <li
          className="rightSideButton githubButton"
        >
          <a
            className="githubButtonLink"
            href="https://github.com/newrelic/opensource-website"
          >
            GitHub
          </a>
        </li>
      </ul>
    </div>
  </div>
  <header
    className="primaryHeaderContainer  "
  >
    <a
      className="primaryHeaderLogo"
      href="/"
    >
      New Relic Open Source
    </a>
    <ul
      className="primaryHeaderNavLinks"
    >
      <li
        className="primaryHeaderNavLinkItem"
      >
        <a
          className="primaryHeaderNavLink"
          href="/external-projects"
        >
          External Projects
        </a>
      </li>
      <li
        className="primaryHeaderNavLinkItem"
      >
        <a
          className="primaryHeaderNavLink"
          href="/explore-projects"
        >
          New Relic Projects
        </a>
      </li>
    </ul>
    <div
      className="primaryHeaderMobileNav "
    >
      <button
        className="mobileMenuButton"
        onClick={[Function]}
      >
        Menu
      </button>
      <ul
        className="primaryHeaderMobileNavLinks"
      >
        <li
          className="primaryHeaderNavLinkItem"
        >
          <a
            className="primaryHeaderNavLink"
            href="/external-projects"
          >
            External Projects
          </a>
        </li>
        <li
          className="primaryHeaderNavLinkItem"
        >
          <a
            className="primaryHeaderNavLink"
            href="/explore-projects"
          >
            New Relic Projects
          </a>
        </li>
      </ul>
    </div>
  </header>
  <main
    className="layout-container-main full-width undefined"
  >
    <div
      className="heroContainer"
    >
      <div
        className="homepageHeroCopy"
      >
        <h2
          className="homepageHeroHeading"
          data-testid="homepageHeroHeading"
        >
          The future of observability is open.
        </h2>
        <p
          className="homepageHeroBodyCopy"
        >
          New Relic ❤️'s open source. We
           
          <a
            href="https://github.com/newrelic/opensource-website"
          >
            built
          </a>
           
          this site to make it easy for 
          <em>
            you
          </em>
           to
           
          <a
            href="/explore-projects"
          >
            explore the projects
          </a>
           we're maintaining as well as our involvement in
           
          <a
            href="/external-projects"
          >
            open standards
          </a>
          . Learn 
          <a
            href="#"
            onClick={[Function]}
          >
            more
          </a>
          .
        </p>
      </div>
      <div
        className="responsiveVideoContainer"
      >
        <div
          className="responsive-video responsiveVideo"
        >
          <iframe
            allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture"
            allowFullScreen={true}
            frameBorder="0"
            height="315"
            src="https://www.youtube.com/embed/7wnav6Fu9T0"
            width="560"
          />
        </div>
      </div>
      <div
        className="homepageHeroVideo"
        onClick={[Function]}
        style={
          Object {
            "backgroundImage": "url(test-file-stub)",
          }
        }
      >
        <img
          alt="close icon"
          className="modalCloseButton"
          onClick={[Function]}
          src="test-file-stub"
        />
        <img
          className="playButton"
          src="test-file-stub"
        />
        <div
          className="iframeContainer"
        >
          <iframe
            allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture; showinfo; modestbranding"
            className="heroVideoIframe"
            frameBorder="0"
            height="562.704471"
            src="https://www.youtube-nocookie.com/embed/7wnav6Fu9T0?showinfo=0&modestbranding=1&rel=0&controls=0"
            width="1000"
          />
        </div>
      </div>
    </div>
    <div
      className="videoModalOverlay"
      onClick={[Function]}
    />
    <div
      className="featuredExternalProjects"
    >
      <h4
        className="featuredExternalProjectsTitle"
      >
        Projects that we support
      </h4>
      <p
        className="featuredExternalProjectsDescription"
      >
        New Relic contributes resources to the development of these projects.
         
        <a
          href="/external-projects"
        >
          View all
        </a>
        .
      </p>
      <div
        className="externalProjectsList"
      />
    </div>
    <div
      className="featuredInternalProjectsContainer"
    >
      <div
        className="featuredInternalProjectsSection"
      >
        <h3
          className="featuredInternalProjectsSectionTitle"
        >
          Explore projects
        </h3>
        <p
          className="featuredInternalProjectsSectionDescription"
        >
          Check out some of the products that we’re developing in open source or 
          <a
            href="/explore-projects"
          >
            view all projects
          </a>
          .
        </p>
      </div>
      <div
        className="featuredInternalProjects"
      />
    </div>
  </main>
  <footer
    className="footerContainer"
  >
    <div
      className="footerContents"
    >
      <div
        className="footerLogoContainer"
      >
        <a
          className="footerLogo "
          href="/"
        >
          New Relic Open Source
        </a>
      </div>
      <div
        className="meta"
      >
        <small
          className="version"
        >
          Version 
          0.5.0
        </small>
        <small
          className="footerLegal"
        >
          Copyright © 
          2020
           New Relic Inc.
        </small>
      </div>
      <ul
        className="footerNavLinks"
      >
        <li
          className="footerNavLinkItem"
        >
          <a
            className="footerNavLink"
            href="/standards"
          >
            Standards
          </a>
        </li>
        <li
          className="footerNavLinkItem"
        >
          <a
            className="footerNavLink"
            href="/external-projects"
          >
            External Projects
          </a>
        </li>
        <li
          className="footerNavLinkItem"
        >
          <a
            className="footerNavLink"
            href="/explore-projects"
          >
            New Relic Projects
          </a>
        </li>
      </ul>
    </div>
  </footer>
</div>
`;
